### PR TITLE
Adding charm upgrade with channel

### PR DESCRIPTION
**Pull request by SimonRichardson**
Allow moving between channels when doing an upgrade charm.

### CODE REVIEW

File: changes.go [hunk 1]
```diff
@@ -293,10 +298,14 @@ func (ch *UpgradeCharmChange) Args() (map[string]interface{}, error) {
...
 // Description implements Change.
 func (ch *UpgradeCharmChange) Description() []string {
-	series := ""
+	var series string
 	if ch.Params.Series != "" {
 		series = " for series " + ch.Params.Series
 	}
+	var channel string
+	if ch.Params.Channel != "" {
+		channel = " for channel " + ch.Params.Channel
```
Comment (achilleasa) [inline changes.go:307]: Does this match the output of `juju deploy --channel`? IIRC we specify the channel in parenthesis. Either way works but we should be consistent.

Comment (SimonRichardson) [reply changes.go:307]: We swap `in` and `from` depending on the context.

Comment (SimonRichardson) [reply changes.go:307]: But yeah it's consistent in form.

**Review by achilleasa — APPROVED** [review]

File: changes.go [hunk 2]
```diff
@@ -31,13 +31,18 @@ type ArchConstraint interface {
 // ConstraintGetter represents a architecture constraint parser.
 type ConstraintGetter func(string) ArchConstraint
 
+// RevisionGetter resolves the revision of a charm from the list of parameters.
+type RevisionGetter func(string, string, string, string) (int, error)
```
Comment (achilleasa) [inline changes.go:35]: Please name the args so there is no mix-up with their order as all of them are strings.

File: handlers.go [hunk 3]
```diff
@@ -230,6 +235,39 @@ func (r *resolver) handleApplications() (map[string]string, error) {
...
+func (r *resolver) allowCharmUpgrade(existingApp *Application, bundleApp *charm.ApplicationSpec, bundleArch string) (bool, error) {
+	// This covers most of v1 charm URL changes, everything else below is to
+	// support channels. Charmstore charms allow channels, but bundles were not
+	// aware of them, with the introduction of Charmhub charms, then we do need
+	// to factor in channels.
+	if existingApp.Charm != bundleApp.Charm {
+		return true, nil
+	}
+	if !r.force && existingApp.Channel != bundleApp.Channel {
+		return false, errors.Errorf("upgrades not supported across channels (existing: %q, requested: %q)", existingApp.Channel, bundleApp.Channel)
```
Comment (achilleasa) [inline handlers.go:247]:
```suggestion
		return false, errors.Errorf("upgrades not supported across channels (existing: %q, requested: %q); use --force to override", existingApp.Channel, bundleApp.Channel)
```

File: handlers.go [hunk 4]
```diff
@@ -230,6 +235,39 @@ func (r *resolver) handleApplications() (map[string]string, error) {
...
+		var err error
+		requestedRev, err = r.revisionGetter(bundleApp.Charm, bundleApp.Series, bundleApp.Channel, bundleArch)
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+	}
+	if requestedRev > existingApp.Revision {
+		return true, nil
+	} else if requestedRev != -1 && requestedRev < existingApp.Revision {
+		return false, errors.Errorf("downgrades are not currently supported")
```
Comment (achilleasa) [inline handlers.go:265]: Technically, you could still do that with V1 with the L244 check. Does it make sense to error here but still allow users to `--force`?

Comment (SimonRichardson) [reply handlers.go:265]: Charmstore allows downgrades, so I don't want to break that

Comment (achilleasa) [reply handlers.go:265]: Right, L244 will take care of that for V1. My question was more about whether we should allow `--force` here to allow a forced downgrade for charmhub as well?

Comment (SimonRichardson) [reply handlers.go:265]: I'll make a note about charmhub and people can raise a bug about it if people encounter it.

**Review by achilleasa — APPROVED** [review]

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**